### DISCUSSION TITLE
Snap package is ready!!!

### DISCUSSION THREAD

**Discussion by soumyaDghosh** [NickvisionApps/Parabolic#94] [General]
Hi, I have made a snap for your app and  I'm ready to maintain it. Would u like to check it out?

**Comment by nlogozzo** [comment]
Looks good! Thanks for maintaining it :)

**Comment by fsobolev** [comment]
@soumyaDghosh 
Not only C# rewrite of the app, but also changes in #58 radically change how the app is built. I have no experience in creating snaps, so I can't help you with any instructions, but I hope I can give you some hints to update snap.
1. There's [dotnet snapcraft plugin](https://snapcraft.io/docs/dotnet-plugin).
2. yt-dlp is now used as python module, not a separate binary, this may require some changes in building.
3. [Blueprint](https://jwestman.pages.gitlab.gnome.org/blueprint-compiler/) and [Just](https://github.com/casey/just) are new build dependencies.
4. `just` is used as some kind of build system. To build self-contained app, you can run `just publish-self-contained PREFIX`. If it doesn't fit your needs and you need another recipe to build snap, you can create a PR to add `publish-snap` recipe into `justfile` (there's a separate recipe for flatpak - `publish-flatpak` - so I see no problem in having another one for snap).

Feel free to ask any questions, I will be happy to help :)

**Comment by soumyaDghosh** [comment]
How should I build the app using `just`, can you give me the command?

**Comment by fsobolev** [comment]
> How should I build the app using `just`, can you give me the command?

```
cd NickvisionTubeConverter.GNOME
just publish-self-contained /usr
just install INSTALL_DIR
```

Replace `/usr` with another prefix if needed. `INSTALL_DIR` should be the path where to install files, use whatever is appropriate for Snap. Examples for other packaging formats: for Flatpak it's `/` (to place everything in the flatpak root), and for AUR it's `$pkgdir` (to place everything in the directory from which the package is created).

**Comment by soumyaDghosh** [comment]
Let's say the `INSTALL_DIR` is `/root/prime/`, then if the prefix is kept as `/usr` then there will be a folder created as `/root/prime/usr`

**Comment by fsobolev** [comment]
> Let's say the `INSTALL_DIR` is `/root/prime/`, then if the prefix is kept as `/usr` then there will be a folder created as `/root/prime/usr`

Right, and the files will be:
> /root/prime/usr/bin/org.nickvision.tubeconverter
/root/prime/usr/share/applications/org.nickvision.tubeconverter.desktop
/root/prime/usr/share/org.nickvision.tubeconverter/org.nickvision.tubeconverter.gresource
/root/prime/usr/lib/org.nickvision.tubeconverter/NickvisionTubeConverter.dll
etc

**Comment by fsobolev** [comment]
@soumyaDghosh We found out that the icon we use for logs button doesn't fit the meaning in Yaru icon theme and probably some other, so in #87 it was hardcoded. Screenshot there shows how the app looks with Yaru icons, and Adwaita icon looks out of style. If you think this is an issue, the suggestion is to replace the icon in Snap package with what was used in old version of the app:
```
sed -i "s/tubeconverter-logs/dialog-information/g" NickvisionTubeConverter.GNOME/Blueprints/download_row.blp
```
This icon will probably look good no matter what Ubuntu flavor a user has installed and what the icon theme is.
But it's up to you, if you think it's okay currently you can leave it as is.

**Comment by soumyaDghosh** [comment]
I will take a look into that. Also, 
Is `liblttng-ust` necessary for your app?

**Comment by nlogozzo** [comment]
> Is liblttng-ust necessary for your app?

I've never heard of that library, nor do we download it anywhere for flatpak, so I'd say no. Unless it's a dependency for `yt-dlp`, but i don't think it is, but not 100% sure.

**Comment by soumyaDghosh** [comment]
Well, biggest issue is, when I'm running it, I'm getting `Unhandled exception.` What can be the reason behind?

**Comment by nlogozzo** [comment]
most likely resource files are not installed correctly...let me wrap the application in a try catch so it will print an error message...

**Comment by nlogozzo** [comment]
try the latest commit in main and see what error message it gives...

**Comment by soumyaDghosh** [comment]
Can I try it later?

**Comment by nlogozzo** [comment]
yeah no rush :)

**Comment by soumyaDghosh** [comment]
Actually on my way to my college and without my lappy... Rushing in😑

**Comment by soumyaDghosh** [comment]
> yeah no rush :)

No, yet I'm getting `Unhandled exception.`

**Reply by nlogozzo** [reply]
So the latest commit in the main branch has the whole running of the application wrapped in a try catch block so you should see an error message if it's coming from Tube Converter.

**Comment by soumyaDghosh** [comment]
I think that's an issue of dotnet. BTW, can you move this to discussion?

**Comment by soumyaDghosh** [comment]
Hello, hope none of you have forgotten me...... If yesss. I'm sorry, I am sorry from my deepest for being so late. But it's never too late right? Here it's, the big picture. Enjoy... And also, love if you'd test it.
<div align="center">
<video controls src="https://user-images.githubusercontent.com/72045785/226995416-11c78648-de2e-4c74-b697-2fdac09140b9.webm" />
</div>

**Reply by fsobolev** [reply]
Yay, it's working, great! 😃 No need to be sorry, we all are volunteering here, doing things when we have time and opportunity :) Thanks for working on snap package.
I didn't find a convenient way to use snap on my main PC with Silverblue, but in a few days I will try it on another device where it will be easier.

**Comment by soumyaDghosh** [comment]
Okay, so, currently I'm uploading this build in the edge channel, until this becomes stable. I'm not gonna make it to the stable channel. To install the latest version, if wished, users have to do
`snap install tube-converter --edge`

**Comment by soumyaDghosh** [comment]
To upload and list this I need to make a reuqest in the snap-store. So, anyone help me writing a better explaination about why I need to read the `/usr/share/zoneinfo-icu/44`. FYI, it's necessary to read it, so that, those messages like `good morning`, `good evening` can be shown.

**Reply by fsobolev** [reply]
Snaps don't have an access to current time by default? 😮 
Okay...

> The application shows a greeting on the start screen that changes depending on the time of day ([code](https://github.com/nlogozzo/NickvisionTubeConverter/blob/snap-2023.3.0/NickvisionTubeConverter.Shared/Controllers/MainWindowController.cs#L94), [screenshot](https://github.com/nlogozzo/NickvisionTubeConverter/blob/main/NickvisionTubeConverter.GNOME/Screenshots/StartScreen.png)), which requires knowing the current time.

**Reply by soumyaDghosh** [reply]
> Snaps don't have an access to current time by default? open_mouth Okay...

They don't even have the access to the internet. Snaps are actually a completely isolated and nonwriteable system. Even using sudo, you can't change things in that. They have a whole different filesystem separated for them, and connect them using interfaces.

**Reply by fsobolev** [reply]
> They don't even have the access to the internet. Snaps are actually a completely isolated and nonwriteable system. Even using sudo, you can't change things in that. They have a whole different filesystem separated for them, and connect them using interfaces.

Yeah, the rest is the same in Flatpak (no internet access by default, separate file tree etc), but no access to system time is... interesting sandbox restriction, huh.

**Reply by soumyaDghosh** [reply]
Hooray.......

https://snapcraft.io/tube-converter


It's there. It also doesn't need the access to system-files

**Reply by fsobolev** [reply]
![tc-snap-wsl](https://user-images.githubusercontent.com/117388856/227803676-1cdee24f-66d1-4310-a221-791a8949715f.png)

Running TC in snap in WSL with openSUSE... 😄 

<img src="https://user-images.githubusercontent.com/117388856/227803813-a86c972b-567b-495b-bf13-aa8dff5ec854.jpg" width=300px>

**Reply by soumyaDghosh** [reply]
Wow... Didn't know that snaps really respects the system theme. Noice....

**Reply by soumyaDghosh** [reply]
Seems like this issue is there also. Can you give me the log which is in the screenshot? Actually there's an running issue within the snaps.
https://forum.snapcraft.io/t/snap-real-home-is-kinda-buggy
Looks like the same issue is here also.

**Reply by fsobolev** [reply]
> Wow... Didn't know that snaps really respects the system theme. Noice....

Not sure what you mean, this is just a usual libadwaita style

> Seems like this issue is there also. Can you give me the log which is in the screenshot? Actually there's an running issue within the snaps.

The issue that I see is that in file chooser home in sidebar opens `$HOME/snap/tube-converter/18`. The lines in screenshot like "...Desktop was removed, reassigning..." appeared because I didn't have xdg-user-dirs installed and so I didn't have all the expected dirs in home. But after I installed it, no errors and in file chooser all sidebar items (except of home) open correct folders. Also, downloading to `$HOME/Videos` works fine.
The full log:

```
(NickvisionTubeConverter.GNOME:1004): Gtk-WARNING **: 16:35:54.456: Unable to acquire the address of the accessibility bus: GDBus.Error:org.freedesktop.DBus.Error.ServiceUnknown: The name org.a11y.Bus was not provided by any .service files. If you are attempting to run GTK without a11y support, GTK_A11Y should be set to 'none'.

(NickvisionTubeConverter.GNOME:1004): Gtk-CRITICAL **: 16:35:54.460: gtk_widget_get_parent: assertion 'GTK_IS_WIDGET (widget)' failed

(NickvisionTubeConverter.GNOME:1004): Gtk-CRITICAL **: 16:35:54.460: gtk_widget_add_css_class: assertion 'GTK_IS_WIDGET (widget)' failed

(NickvisionTubeConverter.GNOME:1004): Gtk-WARNING **: 16:36:03.729: Allocating size to AdwPreferencesGroup 0x563175e81a30 without calling gtk_widget_measure(). How does the code know the size to allocate?

(NickvisionTubeConverter.GNOME:1004): Gtk-CRITICAL **: 16:36:03.735: Allocation height too small. Tried to allocate 424x0, but AdwPreferencesGroup 0x563175e81a30 needs at least 424x70.

(NickvisionTubeConverter.GNOME:1004): Gtk-CRITICAL **: 16:36:03.759: Allocation height too small. Tried to allocate 420x0, but GtkBox 0x563179188f80 needs at least 420x70.

(NickvisionTubeConverter.GNOME:1004): Gdk-WARNING **: 16:36:12.844: Compositor doesn't support moving popups, relying on remapping

(NickvisionTubeConverter.GNOME:1004): Gtk-WARNING **: 16:36:13.527: Failed to load '/home/user/snap/tube-converter/18/.config/gtk-3.0/bookmarks': Error opening file /home/user/snap/tube-converter/18/.config/gtk-3.0/bookmarks: No such file or directory

(NickvisionTubeConverter.GNOME:1004): Gtk-WARNING **: 16:36:13.528: Failed to load '/home/user/snap/tube-converter/18/.config/gtk-3.0/bookmarks': Error opening file /home/user/snap/tube-converter/18/.config/gtk-3.0/bookmarks: No such file or directory

(NickvisionTubeConverter.GNOME:1004): Gtk-WARNING **: 16:36:18.862: Failed to load '/home/user/snap/tube-converter/18/.config/gtk-3.0/bookmarks': Error opening file /home/user/snap/tube-converter/18/.config/gtk-3.0/bookmarks: No such file or directory

(NickvisionTubeConverter.GNOME:1004): Gtk-WARNING **: 16:36:18.863: Failed to load '/home/user/snap/tube-converter/18/.config/gtk-3.0/bookmarks': Error opening file /home/user/snap/tube-converter/18/.config/gtk-3.0/bookmarks: No such file or directory

(NickvisionTubeConverter.GNOME:1004): Gtk-WARNING **: 16:39:25.720: Failed to load '/home/user/snap/tube-converter/18/.config/gtk-3.0/bookmarks': Error opening file /home/user/snap/tube-converter/18/.config/gtk-3.0/bookmarks: No such file or directory

(NickvisionTubeConverter.GNOME:1004): Gtk-WARNING **: 16:39:25.721: Failed to load '/home/user/snap/tube-converter/18/.config/gtk-3.0/bookmarks': Error opening file /home/user/snap/tube-converter/18/.config/gtk-3.0/bookmarks: No such file or directory
```

**Reply by soumyaDghosh** [reply]
That's yaru patch of libadwaita. It's not normal one. I built the libadwaita with the patch by yaru accent colors, and the result is this
![image](https://user-images.githubusercontent.com/72045785/228116796-850ff6d0-29e5-4eea-82f6-2daded34b479.png)

![image](https://user-images.githubusercontent.com/72045785/228116814-0bf06b15-a09d-4483-8f2d-a0044217d5fc.png)

![image](https://user-images.githubusercontent.com/72045785/228116836-9aece9e3-f4b9-4347-b19c-b3bdf311ad94.png)

![image](https://user-images.githubusercontent.com/72045785/228116891-bc14359d-b846-424a-a582-c4c695925609.png)

**Comment by soumyaDghosh** [comment]
I should bundle the `xdg-user-dirs` with the snap
